### PR TITLE
circuit breakers: add retry admission control extension point

### PR DESCRIPTION
**Pull request by AdamEAnderson**
Commit Message: circuit breakers: add retry admission control extension point
Additional Description: Moves the old max_retry and retry_budget circuit breaker options into a new extension point called retry_admission_control. This simplifies the core logic and makes it possible to add alternative implementations for retry limiting.
Risk Level: High -- this involves changes to core code that is run on every HTTP request
Testing: lots of new tests added to cover new functionality, existing tests kept with some modifications
Docs Changes: included much more detailed docs for the two extensions added, as well as changes to the circuit breaking overview docs page.
Release Notes: broke up the change into multiple pieces covering the important portions of the changes.
Platform Specific Features: None
Runtime guard: included as `envoy.reloadable_features.use_retry_admission_control`. Possibly it should be applied a bit wider?
Related to issue: #30205 
Deprecated: Included in release notes. Deprecated the non-extensions based retry circuit breaking fields.
[API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):
- Are the default values going to cause behavior changes for existing users who do not know about the change and have not updated the resources being served by their control plane?
  - No changes to existing behavior from config. Slight changes to runtime override behavior under corner-case scenarios (included in release notes)
  - Existing users of the old API fields should be moved off slowly over a deprecation period via warning messages
- Will users have a way to disable this change if it causes problems?
  - Yes, there's an included runtime guard
- If the feature is modeled as a proto3 scalar, is it plausible that its default value may change in the future? If so, it should be wrapped with a Well-Known Type (WKT), e.g. bool becomes `google.protobuf.BoolValue`.
  - All relevant fields have been wrapped.

### CODE REVIEW

File: source/common/runtime/runtime_features.cc
Comment (AdamEAnderson) [inline source/common/runtime/runtime_features.cc:94]: Added a runtime feature guard to toggle back to the existing `ResourceManager` retry circuit breaking interface. Note that the admission controller will still be created in the cluster info, but will not be used. The feature guard is latched in the router constructor.

This feature guard also serves to decide whether the resource manager or the retry admission control extension gets to write the retry-related circuit breaker stats, which would otherwise be under contention.

The resource manager's retry resource is updated even if this is true, but it isn't consulted prior to allowing a retry.

File: source/common/upstream/upstream_impl.h
Comment (AdamEAnderson) [inline source/common/upstream/upstream_impl.h:1074]: This struct changed a bit to accommodate managing both the resource managers and the admission controls. This is because the stats struct, priority mapping, and runtime key prefix calculations are shared between both.

File: source/extensions/extensions_build_config.bzl
Comment (AdamEAnderson) [inline source/extensions/extensions_build_config.bzl:313]: This extension covers the `max_retries` field by providing a set limit to the number of concurrent retries per cluster and priority combination

File: source/extensions/extensions_build_config.bzl
Comment (AdamEAnderson) [inline source/extensions/extensions_build_config.bzl:314]: This extension covers the `retry_budget` field by allowing a set percentage of active concurrent tries to be retries.

File: source/extensions/extensions_metadata.yaml
Comment (AdamEAnderson) [inline source/extensions/extensions_metadata.yaml:1082]: Someone should check that this metadata makes sense. I'm not sure it does, I just put something that looked sane.

File: source/common/upstream/upstream_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/upstream_impl.cc:2028]: We are deprecating the ResourceManager version of retry circuit breaking by dynamically translating the old config into extension config. This has the upside of moving people off the old config quicker, allowing us to remove the old code before the fields are fully removed.

However, it does mean that given the circuit breaking defaults, the two new extensions are required.

File: test/integration/retry_admission_control/config.proto
Comment (AdamEAnderson) [inline test/integration/retry_admission_control/config.proto:1]: I added a integration test toy extension to be able to validate the basics in a test. This toy extension either allows all retries or denies all retries.

**Comment by AdamEAnderson** [comment]
I realize that this is a quite large PR. If anyone has proposals for splitting it up into smaller pieces, let me know.

**Comment by AdamEAnderson** [comment]
CI appears to be broken right now due to checksum mismatch with a fetched dependency artifact. I will re-run it later once we can merge in a fix.

**Comment by AdamEAnderson** [comment]
cc @tonya11en from the related issue

**Comment by tonya11en** [comment]
I'll take this one on and do the first few review iterations. I'll be back in the office tomorrow. 

Thanks for doing this!

File: changelogs/current.yaml
Comment (markdroth) [reply changelogs/current.yaml:86]: Runtime flags are an Envoy-specific thing, not an xDS thing, so I'll let one of the Envoy maintainers comment on this.  From an xDS perspective, it doesn't seem problematic for runtime flag to control whether a new xDS field is used, since that's not really any different from an older client that doesn't support the new field in the first place.

**Review by tonya11en — CHANGES_REQUESTED** [review]
Thanks for all this work! We'll want to break it up into a few PRs to make the reviews more manageable. It'll also allow you to get things merged incrementally to avoid huge merge conflicts. You can still stack the PRs on top of each other and rebase as necessary.

Here's how I'd suggest doing this based on what you've got here:

**API changes**
Add all the necessary protos and various proto field changes. For this patch, you'll want to add the following to hide the fields until ready:
```
// [#not-implemented-hide:]
// Bleep blorp.
google.protobuf.BoolValue rad = 42;
```

**Implementation of the extensions**
Basically, whatever you have in `extensions/retry/admission_control` and associated tests. Given how significant these changes are, I'd recommend having some tests that just exercise the extension logic.

**Plumbing / Integration**
Whatever is left will probably be fine in this patch. This is where we'll have the conversations around how we'll deprecate the vestigial configs, add the runtime guards, etc.

----

Let me know how that sounds to you? I'll do my best to review/respond within 1 business day.

**Comment by tonya11en** [comment]
Since the API review already happened here, consider simply changing this PR to be the API changes as proposed in my other comment.

File: changelogs/current.yaml
Comment (wbpcode) [reply changelogs/current.yaml:86]: The old static limiter or retry budget should be supported forever by default (although we can moved the code to an extension implementation).

A runtime guard should be unnecessary for this patch. This patch self should ensure only when the new limiter is used, the behavior will be changed.

File: source/common/runtime/runtime_features.cc
Comment (wbpcode) [reply source/common/runtime/runtime_features.cc:94]: Ideally, this patch needn't any runtime guard. The new behaviors/extensions should only be used when the new limiter is set.

**Review by wbpcode — COMMENTED** [review]
Thanks for you contribution. I think the first important thing is to ensure the old behavior will not be effected except the new limiter is set.

**Pull request closed without merging**
